File: Data Structures & Algorithms/rotting-fruit/submission-0.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        ROWS, COLS = len(grid),len(grid[0])
        q = deque()
        fresh, time = 0, 0

        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == 1:
                    fresh += 1
                if grid[r][c] == 2:
                    q.append([r,c])
        
        directions = [
            [1,0],
            [-1,0],
            [0,1],
            [0,-1]
        ]
        while q and fresh > 0:
            for i in range(len(q)):
                r, c = q.popleft()
                for dir_r, dir_c in directions:
                    row, col = r + dir_r, c + dir_c
                    if ( row < 0 or row == ROWS or col < 0 or col == COLS or grid[row][col] != 1):
                        continue
                    grid[row][col] = 2
                    fresh -= 1
                    q.append([row,col])
            time += 1
        return time if fresh == 0 else -1

File: Data Structures & Algorithms/search-2d-matrix/submission-0.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if ( matrix.empty() || matrix[0].empty() ) {
            return false;
        }

        int rows = matrix.size();
        int cols = matrix[0].size();

        int left = 0;
        int right = (rows * cols) - 1;

        while ( left <= right ) {
            int mid = left + ( right - left) / 2;
            int row = mid / cols;
            int col = mid % cols;

            if ( matrix[row][col] == target ) {
                return true;
            } else if ( matrix[row][col] < target ) {
                left = mid + 1; 
            } else if ( matrix[row][col] > target ) {
                right = mid - 1;
            }
        }

        return false;
    }
}; 
